liota/dccs/iotcv2.py

```diff
@@ -0,0 +1,346 @@
+#                                                                             #
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"#
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE  #
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE #
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE  #
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR        #
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF       #
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS   #
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN    #
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)    #
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF     #
+#  THE POSSIBILITY OF SUCH DAMAGE.                                            #
+# ----------------------------------------------------------------------------#
+
+import json
+import logging
+import time
+import threading
+import ConfigParser
+import os
+import Queue
+from time import gmtime, strftime
+from threading import Lock
+
+from liota.dccs.dcc import DataCenterComponent, RegistrationFailure, SetPropertiesFailure, CreateRelationshipFailure
+from liota.lib.protocols.helix_protocol import HelixProtocol
+from liota.entities.metrics.metric import Metric
+from liota.lib.utilities.utility import LiotaConfigPath, getUTCmillis, mkdir, read_liota_config
+from liota.lib.utilities.si_unit import parse_unit
+from liota.entities.metrics.registered_metric import RegisteredMetric
+from liota.entities.registered_entity import RegisteredEntity
+
+log = logging.getLogger(__name__)
+
+timeout = 300
+
+class Request:
+    def __init__(self, guid, operation_id, user_queue):
+        self.guid = guid
+        # assume operation_id will be unique in one process
+        self.operation_id = operation_id
+        self.user_queue = user_queue
+
+class Iotcv2(DataCenterComponent):
+    """ The implementation of VMware IoTC V2 Data Center Component
+
+    """
+
+    def __init__(self, con):
+        log.info("Logging into IOTC V2 DCC")
+        self.comms = con
+        if not self.comms.identity.username:
+            log.error("Username not found")
+            raise ValueError("Username not found")
+        elif not self.comms.identity.password:
+            log.error("Password not found")
+            raise ValueError("Password not found")
+        thread = threading.Thread(target=self.comms.receive)
+        thread.daemon = True
+        # This thread will continuously run in background to receive response or actions from DCC
+        thread.start()
+        time.sleep(0.5)
+        self.counter = 0
+
+        self.recv_msg_queue = self.comms.userdata
+        self.req_ops_lock = Lock()
+        self.req_dict = {}
+        thread = threading.Thread(target=self._dispatch_recvd_msg)
```

Comment:
Can we please name this thread better? Something like message_receive_thread?

Reply:
No, it's not message_receive_thread, but a internal thread for dispatching received messages/responses into each blocking call's internal waiting queue.